pkg/policy/mapstate_test.go

```diff
@@ -1497,153 +1553,235 @@ func (ds *PolicyTestSuite) TestMapState_AccumulateMapChangesOnVisibilityKeys(c *
 		},
 		adds: MapState{},
 		deletes: MapState{
-			testIngressKey(235, 0, 0): testEntry(0, true),
+			testIngressKey(235, 0, 0): allowEntry(0),
+			HttpIngressKey(235):       allowEntryD(12345, visibilityDerivedFrom),
 		},
 	}, {
 		continued: true,
-		name:      "test-6b - Adding an entry that already exists, no adds",
+		name:      "test-2f - Adding an entry that already exists, no adds",
 		args: []args{
-			{cs: csBar, adds: []int{237}, deletes: []int{}, port: 0, proto: 0, ingress: true, redirect: false, deny: true},
+			{cs: csBar, adds: []int{237}, deletes: []int{}, port: 0, proto: 0, ingress: true, redirect: false, deny: false},
 		},
+		visArgs: []visArgs{{
+			redirectPort: 12345,
+			visMeta:      VisibilityMetadata{Ingress: true, Port: 80, Proto: u8proto.TCP},
+		}},
 		state: MapState{
-			testIngressKey(236, 0, 0): testEntry(0, true, csFoo),
-			testIngressKey(237, 0, 0): testEntry(0, true, csBar),
+			testIngressKey(236, 0, 0): allowEntry(0, csFoo).WithDependents(HttpIngressKey(236)),
+			HttpIngressKey(236):       allowEntryD(12345, visibilityDerivedFrom).WithOwners(testIngressKey(236, 0, 0)),
+			testIngressKey(237, 0, 0): allowEntry(0, csBar).WithDependents(HttpIngressKey(237)),
+			HttpIngressKey(237):       allowEntryD(12345, visibilityDerivedFrom).WithOwners(testIngressKey(237, 0, 0)),
 		},
 		adds:    MapState{},
 		deletes: MapState{},
 	}, {
 		continued: false,
-		name:      "test-7a - egress HTTP proxy (setup)",
+		name:      "test-3a - egress HTTP proxy (setup)",
+		setup: MapState{
+			AnyIngressKey():  allowEntry(0),
+			HostIngressKey(): allowEntry(0),
+			HttpEgressKey(0): allowEntry(0),
+		},
 		args: []args{
-			{cs: nil, adds: []int{0}, deletes: []int{}, port: 0, proto: 0, ingress: true, redirect: false, deny: false},
-			{cs: nil, adds: []int{1}, deletes: []int{}, port: 0, proto: 0, ingress: true, redirect: false, deny: false},
 			{cs: csBar, adds: []int{42}, deletes: []int{}, port: 53, proto: 17, ingress: false, redirect: false, deny: false},
 			{cs: csBar, adds: []int{42}, deletes: []int{}, port: 53, proto: 6, ingress: false, redirect: false, deny: false},
 		},
+		visArgs: []visArgs{
+			{
+				redirectPort: 12345,
+				visMeta:      VisibilityMetadata{Ingress: true, Port: 80, Proto: u8proto.TCP},
+			},
+			{
+				redirectPort: 12346,
+				visMeta:      VisibilityMetadata{Ingress: false, Port: 80, Proto: u8proto.TCP},
+			},
+			{
+				redirectPort: 12347,
+				visMeta:      VisibilityMetadata{Ingress: false, Port: 53, Proto: u8proto.UDP},
+			},
+		},
 		state: MapState{
-			AnyIngressKey():     testEntry(0, false, nil),
-			HostIngressKey():    testEntry(0, false, nil),
-			DNSUDPEgressKey(42): testEntry(0, false, csBar),
-			DNSTCPEgressKey(42): testEntry(0, false, csBar),
+			AnyIngressKey():  allowEntry(0),
+			HostIngressKey(): allowEntry(0),
+			// Entry added solely due to visibility annotation has a 'nil' owner
+			HttpIngressKey(0): allowEntryD(12345, visibilityDerivedFrom).WithOwners(nil),
```

Comment:
This test seems a little inconsistent with respect to the ingress visibility policy. In the `setup()`, there's an ingress allow-all and ingress allow-host but no L4 ingress keys. Now this line demonstrates that the resulting state includes an HTTP ingress L4 entry, which I would expect given the visibility policy on redirect port 12345. So this is clearly state that changes when running the logic under test, yet it doesn't end up in the "adds" below.

I understand that this test is only testing the "add" of a couple of egress L3+L4 entries, but if that's the case I'm a bit confused about the role of the ingress visibility policy in this test. Is it to deliberately check that when you add egress allow rules, it will only affect egress and won't touch the ingress rules? But if that's the case, why does the 'state' change to include the ingress visibility policy?

Reply:
The test applies `visArgs` twice, first before applying `args` and then after applying `args`. The `adds` and `deletes` in the test data only account for the latter with the intent of capturing the changes due to `args` only. We could eliminate the first pass of `visArgs` by requiring that `state` already include those keys. First I thought that this would be too hard on the test writer (to manually compute the entries added due to `visArgs`), but now I think I should do that to remove this potential for confusion & but keep the 1st pass of `visArgs` to verify there are no changes (i.e., `state` is consistent).

Reply:
Ended up adding `visAdds` and `visDeletes` to verify the keys added and deleted on the first round of applying `visArgs`. This verifies that those changes are made before applying `args`.

Will squash these changes to the main commit, please review this unit test again, there are no changes elsewhere :-)